### ISSUE TITLE
Assets manager in BDX

### ISSUE THREAD

**Issue by sajjadgrw1** [GoranM/bdx#842]
In libgdx there is assets manager through which we can keep track of what percentage of assets has been loaded like 10% or 99%. Bdx is not using assets manager to load scenes. I am developing game for android platform and i have a scene which is taking long time to load however it is loading fine but current scene is being unresponsive while loading other scene using 
`Bdx.scenes.set(new Scene(0,"Scene"));`
. How can we 
1. Make current scene responsive while other scene is being loaded at background.
2. How to get track of how much percentage of assets/scenes has been loaded like 10% etc.

**Comment by SolarLune** [comment]
Hmm... Well, firstly:

Scenes are loaded immediately currently, as you can tell, on creation. For asynchronous loading, we could use [Threading](http://docs.oracle.com/javase/7/docs/api/java/lang/Thread.html) to create a Thread to load assets for scenes when they're initialized. Not sure if Threads work on all platforms, though. They probably do. 

We could also use LibGDX's asset manager, but we'd still be loading meshes manually, so it wouldn't really work, anyway.

Anyway, to make this a bit easier for people who may not want to spawn Threads, BDX users could enable / disable it with a static switch (i.e. `Bdx.Scene.asynchronousLoading = true`). 

Secondly:

If the first point were implemented, we could fairly easily add a variable or function that could be read that indicates how many assets have been added. If all assets haven't been loaded from a scene, then BDX wouldn't bother updating or rendering anything from it to keep from causing... Interesting problems.

_______

Now, is your scene as optimized as possible? How long does it take to load this scene currently? It might be better to focus on improving the loading time, rather than working on asynchronous loading. Is it because of just the sheer amount of objects, or the number of polygons in your meshes, or the limitations of the device used for testing, or...?

**Comment by SolarLune** [comment]
OK, so I mocked up a basic "this is taking too long, I'm coming back to it later"-style loading process (which can't use Threads for the HTML5 target), but before that:

With a scene of 42 spheres, each with a unique material and ~3K triangles, I ended up with a 48mb scene file. Exporting it from BDX takes roughly 13 seconds. Parsing the JSON file takes around 5.5 seconds, and adding in the data to compose the scene takes around 1.5 seconds.

So the weak points the export from Blender (it takes too long and gives a too heavy result), and parsing JSON (which, again, takes too long). If we could eliminate or minimize these two points, then we might not need asynchronous loading (for at least smaller-scale heavy projects at the current moment).

Either reading directly from the blend file or exporting mesh-specific data would probably be very wise, and might resolve these issues. We should probably keep an eye on #839 and #538 before moving ahead with some sort of asset managing / asynchronous loading.

**Comment by sajjadgrw1** [comment]
Tried to custom build bdx with above commit after fixing multiple issues now facing this issue 
![screen shot 2016-12-02 at 3 02 06 pm](https://cloud.githubusercontent.com/assets/4892234/20830076/b63f83cc-b8a0-11e6-8357-76151b0689fa.png)
even after deleting compiled jars and compiling again using JDK7 then integrating with blender project is running on desktop but above is error for android. Also tried [this](http://stackoverflow.com/questions/37020413/android-dex-cannot-parse-version-52-byte-code) but not fixing it. Is it due to gdx-setup.jar is compiled using JDK8? because that is only jar left others are compiling on runtime using system's JDK which is JDK7 for sure.

**Comment by SolarLune** [comment]
OK, so my previous commit doesn't actually work, it's just something I pushed to save my progress (which is why I didn't try to create a PR for it). I still need to tweak it some more to fix some issues and actually get it to a working condition. If those were the issues you fixed, great. 

Otherwise, it's not done yet, and it's kind of paused at the moment (as I don't know quite what we'll end up doing with import/export). 

________

Secondly, that error message is quite strange, as I didn't do anything with vecmath in my last commit.

Can you compile a vanilla BDX build from master? If not, then I think that would mean that it's not anything to do with my commit, in particular, but rather that you're experiencing other issues with building BDX.

> Is it due to gdx-setup.jar is compiled using JDK8?

I'm unsure of which version of the JDK gdx-setup.jar was compiled with, but I believe @rafcolson uses BDX with JDK7 and exports to Android just fine, so I don't think that's the issue.

________

> Trying another workaround making scene empty and loading models , making GameObjects then adding to scene using assets manager from libgdx is it right approach ? In this way we might have assets manager benefits with good loading time and responsive scenes as well.

Well, the models are manually exported from Blender and imported into BDX - to use the Asset Manager, we'd have to change BDX to export models to a LibGDX-compatible intermediary format, like `.obj` or `.g3db` or something, and then load it in using an Asset Manager with that format.

> Trying another workaround making scene empty and loading models , making GameObjects then adding to scene using assets manager

And I don't quite know what you mean - making the scene empty and then loading the models is what's basically happening. I don't know if you saw my previous post, but loading the models in from `.bdx` files (the JSON parsing process) is the slow part when loading a scene, essentially. In order to make that part "asynchronous", we'd have to write a JSON parser that can be interrupted so that we can continue to update other scenes in BDX. Or make it quicker to export and import models so that we don't need asynchronous loading at the moment.

**Comment by sajjadgrw1** [comment]
Yes you are right , now hacking into bdx source and learning how models are actually being loaded using instantiators and json parsing. Implementing Asynchronous loading would be a great help to reduce load time and increase user interaction. Meanwhile waiting for final commits.
